### PR TITLE
IGNITE-16222 .NET: Thin 3.0: Add RetryPolicy

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Add `IClientRetryPolicy` interface.
* Add predefined policies: `RetryAllPolicy`, `RetryReadPolicy`, `RetryNonePolicy`.
* Add `IgniteClientConfiguration.RetryPolicy` property, defaults to `RetryNonePolicy` (no retries by default).
* Refactor `PooledArrayBufferWriter` to allow sending the same serialized request multiple times.
* Fix table API to use `FailoverSocket` for non-transactional operations to leverage retry/reconnect.

https://cwiki.apache.org/confluence/display/IGNITE/IEP-82+Thin+Client+Retry+Policy

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite/Internal/ClientFailoverSocket.cs [hunk 1]
```diff
@@ -252,5 +268,95 @@ private IEnumerable<IPAddress> GetIps(string host, bool suppressExceptions = fal
...
+            var e = exception;
+
+            while (e != null && !(e is SocketException))
+            {
+                e = e.InnerException;
+            }
+
+            if (e == null)
+            {
+                // Only retry socket exceptions.
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite/Internal/ClientFailoverSocket.cs:292]: Why only socket exceptions?

**Review by isapego — CHANGES_REQUESTED** [review]
Overall looks good, but take a look on few remarks.

File: modules/platforms/dotnet/Apache.Ignite/ClientOperationType.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        /// </summary>
+        TupleUpsert,
+
+        /// <summary>
+        /// Upsert (<see cref="IRecordView{T}.GetAsync"/>).
+        /// </summary>
+        TupleGet,
+
+        /// <summary>
+        /// Upsert (<see cref="IRecordView{T}.UpsertAllAsync"/>).
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/ClientOperationType.cs:48]: Fixed all occurrences.

File: modules/platforms/dotnet/Apache.Ignite/IgniteClientConfiguration.cs [hunk 3]
```diff
@@ -97,5 +99,24 @@ public IgniteClientConfiguration(IgniteClientConfiguration other)
...
+        public IRetryPolicy RetryPolicy { get; set; } = RetryNonePolicy.Instance;
+
+        /// <summary>
+        /// Gets or sets the retry limit. When a request fails due to a connection error,
+        /// Ignite will retry the request if the specified <see cref="RetryPolicy"/> allows it. When this property is
+        /// greater than <c>0</c>, Ignite will limit the number of retries.
+        /// <para />
+        /// Default is <c>0</c>: no limit on retries.
+        /// </summary>
+        public int RetryLimit { get; set; }
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/IgniteClientConfiguration.cs:120]: "Should I retry this" and "how many times should I retry this" are different concerns. Yes, `RetryPolicy` has access to current iteration number and can implement retry limit by itself, but that would mean duplicated retry limit logic in every policy.

File: modules/platforms/dotnet/Apache.Ignite/Internal/ClientFailoverSocket.cs (hunk 1 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/Internal/ClientFailoverSocket.cs:292]: The idea behind the IEP is to retry errors caused by connection issues. `SocketException` is what you get in .NET in those cases.

Comment (isapego) [reply modules/platforms/dotnet/Apache.Ignite/Internal/ClientFailoverSocket.cs:292]: Got it.

File: modules/platforms/dotnet/Apache.Ignite/IgniteClientConfiguration.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite/IgniteClientConfiguration.cs]: Ok, I agree with you. Added `RetryLimitPolicy` class and inherited from it in `RetryAllPolicy`, `RetryReadPolicy`.

**Note:** this is not consistent with Ignite 2.x where `RetryLimit` was introduced before `RetryPolicy`. We can either change 2.x too (deprecate existing property) or keep it as is - not a big deal for a different major version.

[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/dotnet/Apache.Ignite/IgniteClientConfiguration.cs]: I know it is not consistent, but there are so much more inconsistence between 2.x and 3.x that I don't really think this is a problem.

**Review by isapego — APPROVED** [review]

**Pull request merged**